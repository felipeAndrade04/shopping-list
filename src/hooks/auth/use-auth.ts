import { auth } from '@app/config';
import services from '@app/services';
import { LoginParams, RegisterParams } from '@app/services/auth';
import { onAuthStateChanged } from 'firebase/auth';
import { useEffect } from 'react';
import { useAppSelector } from '../use-app-selector';
import { selectAuth, updateLoading, login as loginStore, logout as logoutStore } from '@app/store';
import { useAppDispatch } from '../use-app-dispatch';
import { User } from '@app/models';

export function useAuth() {
  const authState = useAppSelector(selectAuth);
  const dispatch = useAppDispatch();

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (firebaseUser) => {
      if (!firebaseUser) {
        return;
      }

      const user: User = {
        email: firebaseUser.email,
        name: firebaseUser.displayName,
        id: firebaseUser.uid,
      };

      dispatch(loginStore(user));
    });

    return () => {
      unsubscribe();
    };
  }, []);

  async function register(data: RegisterParams) {
    try {
      dispatch(updateLoading(true));

      await services.auth.register(data);

      await login({ email: data.email, password: data.password });
    } catch (error) {
      // const { _message } = error as TypeError;
    } finally {
      dispatch(updateLoading(false));
    }
  }

  async function login(data: LoginParams) {
    try {
      dispatch(updateLoading(true));

      const response = await services.auth.login(data);

      dispatch(loginStore(response));
    } catch (error) {
      // const { _message } = error as TypeError;
    } finally {
      dispatch(updateLoading(false));
    }
  }

  async function logout() {
    try {
      dispatch(updateLoading(true));

      await services.auth.logout();

      dispatch(logoutStore());
    } catch (error) {
      // const { message } = error as TypeError;
    } finally {
      dispatch(updateLoading(false));
    }
  }

  async function forgotPassword(email: string) {
    try {
      dispatch(updateLoading(true));

      await services.auth.forgotPassword(email);
    } catch (error) {
      // const { message } = error as TypeError;
    } finally {
      dispatch(updateLoading(false));
    }
  }

  return {
    register,
    login,
    logout,
    forgotPassword,
    isLoading: authState?.isLoading,
    user: authState?.user,
    isAuthenticated: authState?.isAuthenticated,
  };
}
