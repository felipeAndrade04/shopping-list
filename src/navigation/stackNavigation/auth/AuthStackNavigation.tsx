import React from 'react';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { AuthStackParamList } from './AuthStackNavigation.types';
import { Register, Welcome, Login, ForgotPassword } from '@app/screens';

const Stack = createNativeStackNavigator<AuthStackParamList>();

export function AuthStackNavigator() {
  return (
    <Stack.Navigator
      screenOptions={() => ({
        headerShown: false,
      })}
    >
      <Stack.Screen name="Welcome" component={Welcome} />
      <Stack.Screen name="Login" component={Login} />
      <Stack.Screen name="Register" component={Register} />
      <Stack.Screen name="ForgotPassword" component={ForgotPassword} />
    </Stack.Navigator>
  );
}
