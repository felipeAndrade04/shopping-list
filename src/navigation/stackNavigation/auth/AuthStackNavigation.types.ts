import { NativeStackNavigationProp } from '@react-navigation/native-stack';

export type AuthStackParamList = {
  Welcome: undefined;
  Login: undefined;
  Register: undefined;
  ForgotPassword: undefined;
};

export type AuthStackNavigationProps = NativeStackNavigationProp<AuthStackParamList>;
