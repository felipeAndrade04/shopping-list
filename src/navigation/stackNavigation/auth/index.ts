export * from './AuthStackNavigation';
export * from './AuthStackNavigation.types';
