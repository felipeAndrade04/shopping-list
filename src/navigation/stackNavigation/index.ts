export * from './auth';
export * from './MainStackNavigation';
