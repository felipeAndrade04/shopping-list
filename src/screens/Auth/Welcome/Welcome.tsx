import React from 'react';
import { StatusBar } from 'react-native';
import * as S from './Welcome.styles';
import { Button, Message, Spacer } from '@app/components';
import { AuthStackParamList } from '@app/navigation/stackNavigation/auth';
import Logo from '@app/assets/images/largeLogo.svg';
import { WelcomeProps } from './Welcome.types';

export function Welcome({ navigation }: WelcomeProps) {
  function onNavigation(name: keyof AuthStackParamList) {
    navigation.navigate(name);
  }

  return (
    <S.Container>
      <StatusBar barStyle={'dark-content'} />

      <S.LogoContainer>
        <Logo />
      </S.LogoContainer>

      <Message
        showImage={false}
        title="Olá!"
        description="Comece agora a organizar suas idas ao mercado com lista de compras personalizadas e compartilháveis!"
      />
      <S.Buttons>
        <Button onPress={() => onNavigation('Login')}>Entrar</Button>
        <Spacer dimesion={16} />
        <Button variant="outline" onPress={() => onNavigation('Register')}>
          Criar uma conta
        </Button>
      </S.Buttons>
    </S.Container>
  );
}
