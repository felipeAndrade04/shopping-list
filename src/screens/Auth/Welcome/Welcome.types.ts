import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { AuthStackParamList } from '@app/navigation/stackNavigation/auth';

export type WelcomeProps = NativeStackScreenProps<AuthStackParamList, 'Welcome'>;
