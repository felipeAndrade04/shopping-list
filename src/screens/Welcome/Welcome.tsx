import { Button, Message } from '@app/components';
import React from 'react';
import { StatusBar } from 'react-native';
import * as S from './Welcome.styles';

export function Welcome() {
  return (
    <S.Container>
      <StatusBar barStyle={'dark-content'} />
      <Message
        title="Olá!"
        description="Comece agora a organizar suas idas ao mercado com lista de compras personalizadas e compartilháveis!"
      />
      <S.Buttons>
        <Button>Entrar</Button>
        <Button variant="outline" margin="24px 0 0 0">
          Criar uma conta
        </Button>
      </S.Buttons>
    </S.Container>
  );
}
