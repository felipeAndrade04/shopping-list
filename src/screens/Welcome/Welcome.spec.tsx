import React, { fireEvent, render } from '@testing-library/react-native';
import { Welcome } from '../Welcome';
import { AuthStackNavigationProps } from '@app/navigation/stackNavigation/auth';
import { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import * as theme from '@app/theme';

describe('Welcome page', () => {
  const navigation = {
    ...jest.requireMock('@app/navigation/stackNavigation/auth'),
    navigate: jest.fn(),
  } as AuthStackNavigationProps;

  const wrapper = ({ children }: { children: ReactNode }) => (
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  );

  it('Should render correctly', () => {
    const tree = render(<Welcome navigation={navigation}></Welcome>, { wrapper }).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should navigate to login page', () => {
    const { queryByText } = render(<Welcome navigation={navigation}></Welcome>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);
    fireEvent.press(button);

    expect(navigation.navigate).toHaveBeenCalledWith('Login');
  });

  it('Should navigate to register', () => {
    const { queryByText } = render(<Welcome navigation={navigation}></Welcome>, {
      wrapper,
    });

    const button = queryByText(/criar uma conta/i);
    fireEvent.press(button);

    expect(navigation.navigate).toHaveBeenCalledWith('Register');
  });
});
