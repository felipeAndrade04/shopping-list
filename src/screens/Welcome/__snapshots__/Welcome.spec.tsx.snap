// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome page Should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#EEEEEE",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "paddingBottom": 24,
        "paddingLeft": 24,
        "paddingRight": 24,
        "paddingTop": 24,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "height": 200,
            "marginBottom": 50,
            "marginLeft": "auto",
            "marginRight": "auto",
            "marginTop": 0,
            "width": 240,
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#0D0A0B",
            "fontFamily": "Poppins-Medium",
            "fontSize": 24,
          },
        ]
      }
    >
      Olá!
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#0D0A0B",
            "fontFamily": "Poppins-Regular",
            "fontSize": 16,
            "marginTop": 14,
          },
        ]
      }
    >
      Comece agora a organizar suas idas ao mercado com lista de compras personalizadas e compartilháveis!
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": 48,
          "width": "100%",
        },
      ]
    }
  >
    <RNGestureHandlerButton
      activeOpacity={0.105}
      collapsable={false}
      color="#009Cf7"
      delayLongPress={600}
      handlerTag={1}
      handlerType="NativeViewGestureHandler"
      onActiveStateChange={[Function]}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#009Cf7",
          "borderColor": "black",
          "borderRadius": 6,
          "borderStyle": "solid",
          "borderWidth": 0,
          "display": "flex",
          "height": 56,
          "justifyContent": "center",
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
      underlayColor="black"
      variant="solid"
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "black",
            "borderBottomLeftRadius": undefined,
            "borderBottomRightRadius": undefined,
            "borderRadius": 6,
            "borderTopLeftRadius": undefined,
            "borderTopRightRadius": undefined,
            "bottom": 0,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <Text
        color="#009Cf7"
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "700",
            },
          ]
        }
        variant="solid"
      >
        Entrar
      </Text>
    </RNGestureHandlerButton>
    <View
      dimesion={16}
      style={
        Array [
          Object {
            "height": 16,
            "width": 16,
          },
        ]
      }
    />
    <RNGestureHandlerButton
      activeOpacity={0.105}
      collapsable={false}
      color="#009Cf7"
      delayLongPress={600}
      handlerTag={2}
      handlerType="NativeViewGestureHandler"
      onActiveStateChange={[Function]}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#009Cf7",
          "borderRadius": 6,
          "borderStyle": "solid",
          "borderWidth": 2,
          "display": "flex",
          "height": 56,
          "justifyContent": "center",
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
      underlayColor="black"
      variant="outline"
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "black",
            "borderBottomLeftRadius": undefined,
            "borderBottomRightRadius": undefined,
            "borderRadius": 6,
            "borderTopLeftRadius": undefined,
            "borderTopRightRadius": undefined,
            "bottom": 0,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <Text
        color="#009Cf7"
        style={
          Array [
            Object {
              "color": "#009Cf7",
              "fontSize": 18,
              "fontWeight": "700",
            },
          ]
        }
        variant="outline"
      >
        Criar uma conta
      </Text>
    </RNGestureHandlerButton>
  </View>
</View>
`;
